Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducer.swift

```diff
@@ -0,0 +1,1064 @@
+    @usableFromInline
+    /* fileprivate */ internal final class Storage: @unchecked Sendable {
+        /// The lock that protects our state.
+        @usableFromInline
+        /* private */ internal let lock = Lock()
+        /// The state machine.
+        @usableFromInline
+        /* private */ internal var stateMachine: StateMachine
+        /// The delegate.
+        @usableFromInline
+        /* private */ internal var delegate: Delegate?
+
+        @usableFromInline
+        /* fileprivate */ internal init(
+            backPressureStrategy: Strategy,
+            delegate: Delegate
+        ) {
+            self.stateMachine = .init(backPressureStrategy: backPressureStrategy)
+            self.delegate = delegate
+        }
+
+        @inlinable
+        /* fileprivate */ internal func sequenceDeinitialized() {
+            let delegate: Delegate? = self.lock.withLock {
+                let action = self.stateMachine.sequenceDeinitialized()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self.delegate
+                    self.delegate = nil
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate()
+        }
+
+        @inlinable
+        /* fileprivate */ internal func iteratorInitialized() {
+            self.lock.withLock {
+                let action = self.stateMachine.iteratorInitialized()
+
+                switch action {
+                case .none:
+                    return
+                }
+            }
+        }
+
+        @inlinable
+        /* fileprivate */ internal func iteratorDeinitialized() {
+            let delegate: Delegate? = self.lock.withLock {
+                let action = self.stateMachine.iteratorDeinitialized()
+
+                switch action {
+                case .callDidTerminate:
+                    let delegate = self.delegate
+                    self.delegate = nil
+
+                    return delegate
+
+                case .none:
+                    return nil
+                }
+            }
+
+            delegate?.didTerminate()
+        }
+
+        @inlinable
+        /* fileprivate */ internal func yield<S: Sequence>(_ sequence: S) -> Source.YieldResult where S.Element == Element {
+            self.lock.withLock {
+                let action = self.stateMachine.yield(sequence)
+
+                switch action {
+                case .returnProduceMore:
+                    return .produceMore
+
+                case .returnStopProducing:
+                    return .stopProducing
+
+                case .returnDropped:
+                    return .dropped
+
+                case .resumeContinuationAndReturnProduceMore(let continuation, let element):
+                    // It is safe to resume the continuation while holding the lock
```

Comment:
This is somewhat documented on the `resume` method:

> After resume enqueues the task, control immediately returns to the caller. The task continues executing when its executor is able to reschedule it.

What this roughly translates to being is: `dispatchQueue.async {}`. It is safe to hold a lock across this, since the enqueued job will just wait on the next `lock.lock()`. Furthermore, we are releasing the lock right after resuming so there really shouldn't be too much waiting.

Reply:
Yes, this is safe 👍